lib/plugin.js

```diff
@@ -50,4 +50,36 @@ function resolve (plugins, emitter) {
   return modules
 }
 
-exports.resolve = resolve
+const pluginInstances = new Map()
+
+function createInstantiatePlugin (injector) {
+  const emitter = injector.get('emitter')
+  return function instantiatePlugin (kind, name) {
+    if (pluginInstances.has(name)) {
+      return pluginInstances.get(name)
+    }
+
+    let p
+    try {
+      p = injector.get(`${kind}:${name}`)
+      if (!p) {
+        log.error(`Failed to instantiate ${kind} ${name}`)
+        emitter.emit('load_error', `${kind}`, name)
+      }
+    } catch (e) {
+      if (e.message.includes(`No provider for "${kind}:${name}"`)) {
+        log.error(`Can not load "${name}", it is not registered!\n  Perhaps you are missing some plugin?`)
+      } else {
+        log.error(`Can not load "${name}"!\n  ` + e.stack)
+      }
+      emitter.emit('load_error', `${kind}`, name)
+    }
+
+    pluginInstances.set(name, p)
+    return p
+  }
+}
```

Comment:
Which module?

Reply:
https://www.npmjs.com/package/di

Reply:
I agree that the cache is completely redundant and I will remove that part. The rest is error handling which could be in `di` but is not yet.

Reply:
I wonder if the cache was there to prevent multiple attempts to instantiate a missing preprocessor and consequently the same error being logged a lot of times 🤔